community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/spi/RepeatableReadQueryContext.scala

```diff
@@ -113,11 +114,51 @@ final class RepeatableReadQueryContext(inner: QueryContext, locker: Locker) exte
     locker.acquireLock(relationshipOps.getByInnerId(id))
   }
 
-  private def lockAll[T <: PropertyContainer](iter: Iterator[T]): Iterator[T] = iter.map {
-    item =>
-      locker.acquireLock(item)
-      item
+
+  private def lockAllNodesAndFilter(inner: Iterator[Node])(f: Node => Boolean): Iterator[Node] = new Iterator[Node] {
+    var nextNode: Option[Node] = null
+
+    private def fetchNext() =
+      if (inner.isEmpty) {
+        nextNode = None
+      }
+      else {
+        var found: Option[Node] = None
+        while (inner.hasNext && found.nonEmpty) {
+
+          val temp: Node = inner.next()
+          locker.acquireLock(temp)
+          if (!nodeOps.isDeleted(temp) && f(temp)) {
+            found = Some(temp)
+          }
+        }
+        if (found.isEmpty)
+          nextNode = None
+      }
+
+    override def hasNext = {
+      if (nextNode == null) fetchNext()
+      nextNode.nonEmpty
+    }
+
+    override def next() = {
+      if (nextNode == null) fetchNext()
+
+      nextNode match {
+        case None => Iterator.empty.next() // Throw
```

Comment:
This seems a bit awkward. Why not explicitly throw an exception with a nice error message?